lib/healthcheck/healthcheck-control.c

```diff
@@ -28,6 +28,27 @@
 #include "control/control-connection.h"
 #include "control/control-commands.h"
 #include "stats/stats-prometheus.h"
+#include "stats/stats-registry.h"
+#include "stats/stats-cluster-logpipe.h"
+#include "stats/stats-counter.h"
+
+static inline void
+_append_internal_queue_events_if_exists(GString *reply)
+{
+  StatsClusterKey sc_key;
+  stats_cluster_logpipe_key_set(&sc_key, "internal_events_total", NULL, 0);
+  stats_cluster_logpipe_key_add_legacy_alias(&sc_key, SCS_GLOBAL, "internal_source", NULL );
+
+  stats_lock();
+  StatsCounterItem *internal_queued = stats_get_counter(&sc_key, SC_TYPE_QUEUED);
+  stats_unlock();
+
+  if (!internal_queued)
+    return;
+
+  g_string_append_printf(reply, PROMETHEUS_METRIC_PREFIX
+                         "internal_queue_events %"G_GSIZE_FORMAT"\n", stats_counter_get(internal_queued));
+}
```

Comment:
It is a bit confusing to see the value of this counter in the output, but it not being calculated into the healthcheck's result.

Can we expose some functions from `afinter`, like `get_queue_length()` and `is_queue_filled()` or something like that, and include that in the health calculation?

---

Also, the latency related metrics are exactly the same in the `stats prometheus` output and `healthcheck` output, maybe we should do the same for this, too.

Reply:
The healthcheck unit is for doing real calculations. Since there is nothing to "check", it is here in the report.

The stats prometheus output contains processed/queued/dropped counters. Only the queue length is interesthing in a healthcheck report in my opinion, so I changed the name.

Reply:
I though that if the queue is stuck (queue_length == queue_limit), it means syslog-ng is not healthy.

Reply:
The internal queue's capacity is currently not exposed in stats Prometheus or here.
We can add a metric to both locations, but that would be a separate key, not a label.

I think syslog-ng is unhealthy if queue_length>E for more than a few seconds, where E is near 0 depending on the cost of the internal source's pipeline (destination queues should hold those messages, not the source).

I was thinking about showing the dropped counter instead, but I dismissed the idea as healthcheck should be about the current state of syslog-ng, not past states.

What do you think?

Reply:
I agree that if queue_length is more than 0 for more then a few seconds, it means syslog-ng is unhealthy. And I think it is hard to measure right now, as we only do one measurement.

But I would also say that if queue_length reaches queue_limit (which is much larger than the E you have mentioned) for even a split second, it also means that syslog-ng is unhealthy. This is a lot easier to decide, as we only need one measurement.

(IMO this is similar to cooking meat, where you can be sure that it is not raw if it has X temperature for multiple minutes/hours or if it has a higher Y temperature for a second. :D)

If I were to implement this, I would expose the two functions from afinter like I mentioned (fortunately it is in `lib/`), and I would not operate on its stats counters, but use the two functions to create and print the metric of its queue status and to decide its health.